src/background/utils/requests.js

```diff
@@ -107,52 +108,61 @@ const HeaderInjector = (() => {
   };
 })();
 
-export function getRequestId() {
+export function getRequestId(eventsToNotify = []) {
+  eventsToNotify.push('loadend');
   const id = getUniqId();
   requests[id] = {
     id,
+    eventsToNotify,
     xhr: new XMLHttpRequest(),
   };
   return id;
 }
 
 function xhrCallbackWrapper(req) {
-  let lastPromise = Promise.resolve();
-  const { xhr } = req;
-  return (evt) => {
-    const res = {
-      id: req.id,
-      type: evt.type,
-      resType: xhr.responseType,
-      contentType: xhr.getResponseHeader('Content-Type') || 'application/octet-stream',
-    };
+  let lastPromise;
+  let blobUrl;
+  const { id, xhr } = req;
+  return async (evt) => {
+    if (evt.type === 'loadend') clearRequest(req);
+    if (!req.cb) return;
     const data = {
       finalUrl: xhr.responseURL,
       readyState: xhr.readyState,
       responseHeaders: req.responseHeaders || xhr.getAllResponseHeaders(),
       status: xhr.status,
       statusText: xhr.statusText,
+      response: xhr.responseType === 'blob' && xhr.response ? getBlobUrl(req) : xhr.response,
+      ...tryOrIgnore(objectPick, xhr, ['responseText']),
+      ...('loaded' in evt) && objectPick(evt, ['lengthComputable', 'loaded', 'total']),
     };
-    res.data = data;
+    await lastPromise;
+    lastPromise = req.cb({
+      id,
+      data,
+      contentType: xhr.getResponseHeader('Content-Type') || 'application/octet-stream',
+      resType: xhr.responseType,
+      type: evt.type,
+    });
```

Comment:
This is not equivalent to the current behavior.

Take this example:
- While processing *callback1*, there are two new tasks being sceduled, say *callback2* and *callback3*, both waiting for `lastPromise`, a promise for *callback1*.
- After *callback1* is resolved, `lastPromise` will be set twice immediately, once by *callback2* and once by *callback3*.
- As a result, *callback2* and *callback3* are executed concurrently, not as expected.

So `await` is not supposed to be used here because our real purpose is to update `lastPromise` immediately and we don't care about when it is resolved.

Reply:
Indeed.